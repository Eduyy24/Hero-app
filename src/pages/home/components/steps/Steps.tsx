import { ProgressBar } from "../../../../components/progress-bar/ProgressBar";
import { WrapperSteps } from "../../../../components/wrapper-steps/WrapperSteps";
import logoHero from "../../../../assets/superhero.png";
import styles from "./Steps.module.css";
import { FunctionComponent, useState } from "react";
import Components from "../../../../components/Components";
import { GeneralButton } from "../../../../components/general-button/GeneralButton";
import { NAME_BUTTON_STEPS } from "../../../../utils/strings";
import { useForm } from "react-hook-form";
import { useDispatch, useSelector } from "react-redux";
import {
  GeneralState,
  setDataSell,
} from "../../../../redux/slices/generalSlice";
import { useNavigate } from "react-router-dom";
import usePagesData from "../../../../hooks/usePagesData";
import { Modal } from "../../../../components/modal/Modal";

type Props = {
  keyForm: string;
};

export const Steps: FunctionComponent<Props> = ({ keyForm }) => {
  const { getPageForKey, getAllPagesData, getPageForOrder } = usePagesData();
  const dispatch = useDispatch();
  const navigate = useNavigate();
  const [openModal, setOpenModal] = useState(false)


  const allPages = getAllPagesData();
  const currentPage = getPageForKey(keyForm || "");
  const nextPath = getPageForOrder((currentPage?.order || 0) + 1)?.key;

  const fieldsKey = allPages.reduce((acum, page) => {
    const keys = page.fields.map((field) => field.key) as never[];
    return [...acum, ...keys];
  }, []);

  const dataSet = useSelector(
    (state: { general: GeneralState }) => state.general.dataSell
  );

  const defaultValues = fieldsKey?.reduce((acum, value) => {
    return { ...acum, [value]: dataSet[value] ?? "" };
  }, {});

  const form = useForm({
    defaultValues,
  });

  const onSubmit = (data: any) => {
    dispatch(setDataSell(data));
    navigate(`/${nextPath ?? "resume"}`);
  };

  const onCloseModal = () => setOpenModal(false)

  const onClickOpenModal = () => setOpenModal(true)

  return (
    <WrapperSteps>
      <div data-testid='container-steps'>
        <div className={styles.containerResumeBtn}>
          <p onClick={onClickOpenModal} className={styles.resumeBtn}>Resumen</p>
        </div>
        <img className={styles.logo} src={logoHero} alt="logo hero" />
        <ProgressBar
          value={currentPage?.order || 0}
          max={allPages.length || 0}
        />
        <form onSubmit={form.handleSubmit(onSubmit)}>
          {currentPage?.fields.map((field) =>
            Components(field, form)
          )}
          <div className={styles.containerButton}>
            <GeneralButton name={NAME_BUTTON_STEPS} />
          </div>
        </form>
        <Modal onCloseModal={onCloseModal} open={openModal}>
          <p>Este es un Resumen</p>
        </Modal>
      </div>
    </WrapperSteps>
  );
};
