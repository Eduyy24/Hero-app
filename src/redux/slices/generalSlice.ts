import { createSlice } from '@reduxjs/toolkit'

export interface GeneralState {
  pagesData: PageData[];
  isFullData: boolean;
  dataSell: {};
}

const initialState = {
  pagesData: [],
  isFullData: false,
  dataSell: {}
} as GeneralState

export const generalSlice = createSlice({
  name: 'general',
  initialState,
  reducers: {
    setPagesData: (state, action) => {
      state.pagesData = action.payload
      state.isFullData = true;
    },
    setDataSell: (state, action) => {
      state.dataSell = {...state.dataSell, ...action.payload}
    }
  },
})

export const { setPagesData, setDataSell } = generalSlice.actions

export default generalSlice.reducer