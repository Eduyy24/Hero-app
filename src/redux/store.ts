import { configureStore } from '@reduxjs/toolkit'
import thunk from 'redux-thunk';
import storage from 'redux-persist/lib/storage'

import { persistReducer, persistStore } from 'redux-persist';
import generalReducer from './slices/generalSlice'


const persistConfig = {
  key: 'root',
  storage,
}

const persistedReducer = persistReducer(persistConfig, generalReducer)

const store = configureStore({
  reducer: {
    general: persistedReducer
  },
  devTools: process.env.NODE_ENV !== 'production',
  middleware:[thunk]
})

const persistor = persistStore(store);

export { store, persistor };
