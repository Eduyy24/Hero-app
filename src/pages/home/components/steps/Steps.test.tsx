import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { Provider } from "react-redux";
import { setPagesData } from "../../../../redux/slices/generalSlice";
import { store } from "../../../../redux/store";
import {Steps} from './Steps';


describe("<Steps />", () => {

  test("<InputData> validate useForm hook with JSON config for the component", async () => {
    const validEmail = 'Eduardo@edu.com'
    const inValidEmail = 'Eduardo@edu'

    store.dispatch(setPagesData([
      {
        order: 1,
        key: 'step-1',
        fields: [
          {
            type: 'string',
            component: 'inputText',
            key: 'email',
            label: 'Correo electrómnico',
            rules: {
              required: { value: true, message: 'El campo es obligatorio' },
              pattern: { value: /^[^\s@]+@[^\s@]+\.[^\s@]+$/, message: 'Correo inválido' },
            }
          }
        ]
      },
    ]))

    render(
      <Provider store={store}>
        <Steps keyForm="step-1" />
      </Provider>
    );
    const inputs = screen.getAllByTestId('inputText')
    const content = screen.getByTestId('container-steps')
    const btnContinue = screen.getByText('Continuar')

    expect(content).toMatchSnapshot();
    expect(inputs.length).toEqual(1) // 1 input in de document

    fireEvent.submit(btnContinue);

    await waitFor(() => {
      // valido la aplicación de las reglas y la aparición del mensaje de error segun el JSON
      expect(screen.getByText('El campo es obligatorio')).toBeInTheDocument();
    })

    fireEvent.input(inputs[0], {target: {
      value: inValidEmail
    }})

    await waitFor(() => {
      // valido la aplicación de las reglas y la aparición del mensaje de error segun el JSON
      expect(screen.getByText('Correo inválido')).toBeInTheDocument();
    })

    fireEvent.input(inputs[0], {target: {
      value: validEmail
    }})

    fireEvent.submit(btnContinue);

    await waitFor(() => {
      // valido el correcto funcionamiento del acction al guardar en el dataSell
      const dataSell = store.getState().general.dataSell as {email: string};
      expect(dataSell?.email).toBe(validEmail)
    })

    await waitFor(() => {
      // Valido el ocultamiento del mensaje de error una vez diligenciado el campo correctamente
      expect(screen.queryByText('El campo es obligatorio')).not.toBeInTheDocument()
    })
  });

  test("<InputSelect> validate useForm hook with JSON config for the component", async () => {
    store.dispatch(setPagesData([
      {
        order: 2,
        key: 'step-2',
        fields: [
          {
            type: 'string',
            component: 'inputSelect',
            key: 'girlFriend',
            label: 'Tiene novia?',
            options: ['SI', 'NO'],
            rules: {
              required: { value: true, message: 'El campo es obligatorio' },
            }
          }
        ]
      },
    ]))

    render(
      <Provider store={store}>
        <Steps keyForm="step-2" />
      </Provider>
    );
    const inputs = screen.getAllByTestId('inputSelect')
    const content = screen.getByTestId('container-steps')
    const btnContinue = screen.getByText('Continuar')

    expect(content).toMatchSnapshot();
    expect(inputs.length).toEqual(1) // 1 input in de document

    fireEvent.submit(btnContinue);

    await waitFor(() => {
      // valido la aplicación de las reglas y la aparición del mensaje de error segun el JSON
      expect(screen.getByText('El campo es obligatorio')).toBeInTheDocument();
    })

    userEvent.selectOptions(inputs[0], 'SI');

    fireEvent.submit(btnContinue);

    await waitFor(() => {
      // valido el correcto funcionamiento del acction al guardar en el dataSell
      const dataSell = store.getState().general.dataSell as {girlFriend: string};
      expect(dataSell?.girlFriend).toBe('SI')
    })

    await waitFor(() => {
      // Valido el ocultamiento del mensaje de error una vez diligenciado el campo correctamente
      expect(screen.queryByText('El campo es obligatorio')).not.toBeInTheDocument()
    })
  })

  test("<InputFile> validate useForm hook with JSON config for the component", async () => {
    // implement
  })

  test("<InputCheckbox> validate useForm hook with JSON config for the component", async () => {
    // implement
  })
});