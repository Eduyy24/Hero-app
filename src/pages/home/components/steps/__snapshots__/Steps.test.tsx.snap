// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Steps /> <InputData> validate useForm hook with JSON config for the component 1`] = `
<div
  data-testid="container-steps"
>
  <img
    alt="logo hero"
    class="logo"
    src="superhero.png"
  />
  <div>
    <progress
      class="progressBar"
      data-testid="progress"
      max="1"
      value="1"
    />
    <p
      class="count"
    >
      1
       / 
      1
    </p>
  </div>
  <form>
    <div
      class="containerInput"
    >
      <label
        class="label"
        for="input"
      >
        Correo electr√≥mnico
      </label>
      <input
        class="inputText"
        data-testid="inputText"
        id="input"
        name="email"
        value=""
      />
    </div>
    <div
      class="containerButton"
    >
      <button
        class="generalButton"
        name="Continuar"
      >
        Continuar
      </button>
    </div>
  </form>
</div>
`;

exports[`<Steps /> <InputSelect> validate useForm hook with JSON config for the component 1`] = `
<div
  data-testid="container-steps"
>
  <img
    alt="logo hero"
    class="logo"
    src="superhero.png"
  />
  <div>
    <progress
      class="progressBar"
      data-testid="progress"
      max="1"
      value="2"
    />
    <p
      class="count"
    >
      2
       / 
      1
    </p>
  </div>
  <form>
    <div
      class="containerSelect"
    >
      <label
        class="label"
      >
        Tiene novia?
      </label>
      <select
        class="selectInput"
        data-testid="inputSelect"
        name="girlFriend"
      >
        <option
          value=""
        >
          Seleccione
        </option>
        <option
          value="SI"
        >
          SI
        </option>
        <option
          value="NO"
        >
          NO
        </option>
      </select>
    </div>
    <div
      class="containerButton"
    >
      <button
        class="generalButton"
        name="Continuar"
      >
        Continuar
      </button>
    </div>
  </form>
</div>
`;
